src/dict.c

```diff
@@ -1003,7 +1011,7 @@ static unsigned long rev(unsigned long v) {
 unsigned long dictScan(dict *d,
                        unsigned long v,
                        dictScanFunction *fn,
-                       dictScanBucketFunction* bucketfn,
+                       dictDefragAllocFunction *defragallocfn,
```

Comment:
The type of the callback here is `void*` to `void*`, which is sort of generic and optional. Making it global would make it more coupled with defrag IMHO. Perhaps it's just the callback name that's ugly?

I thought about taking a 'realloc' callback with the same type signature as realloc and which behaves more like realloc (returning the same pointer rather than NULL). Then defrag can pass a wrapper to activeDefragAlloc wrapped in a realloc signature. Not sure if this is more generic. WDYT?

Reply:
the callback would still be specific for defrag purposes.
i.e. in the old code it was a generic per-bucket callback and the defrag part was implemented in the caller, so dict.c was generic.
now dict.c has a defrag feature, and the way to enable it is to pass that pointer. so renaming it won't do much help.

maybe it'll look better if we have two scan interfaces `dictScan` where we'll remove that argument, and `dictScanAndDefrag` (or just dictDefrag).
the implementation can be put in dictScanGeneric, and the two above will just be thin wrappers.
then when we have a dictDefrag function, the boolean or pointer to activeDefragAlloc isn't needed (interface wise).
the generic function does take a boolean, and has explicit calls to activeDefragAlloc which we'll put in server.h

i'll stress the point again, dict.c has a defrag feature now, we can't make that pointer a generic feature, and there's no need to sugarcoat it..

Reply:
> maybe it'll look better if we have two scan interfaces `dictScan` where we'll remove that argument, and `dictScanAndDefrag` (or just dictDefrag).
> the implementation can be put in dictScanGeneric, and the two above will just be thin wrappers.

OK, I can do this.

Reply:
Done (but i didn't add dictScanGeneric. dictScanDefrag is the generic one with both scan and defrag).